File: arrays/medium/53MaximumSubArray/solution.cpp
// solution.cpp
#include <bits/stdc++.h>
using namespace std;
int maxSubArray(vector<int> &nums) {
  int cur = nums[0], best = nums[0];
  for (int i = 1; i < nums.size(); ++i) {
    cur = (cur < 0) ? nums[i] : cur + nums[i];
    best = max(best, cur);
  }
  return best;
}
int main() {
  vector<int> v = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
  cout << maxSubArray(v) << "\n";
}

File: arrays/medium/53MaximumSubArray/solution.rs
// solution.rs
fn max_sub_array(nums: &[i32]) -> i32 {
    let mut cur = nums[0];
    let mut best = nums[0];
    for &x in &nums[1..] {
        if cur < 0 {
            cur = x
        } else {
            cur += x
        }
        if cur > best {
            best = cur
        }
    }
    best
}
fn main() {
    println!("{}", max_sub_array(&[-2, 1, -3, 4, -1, 2, 1, -5, 4]));
}
